modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateIndexNameProcessor.java

```diff
@@ -94,7 +98,7 @@ public void execute(IngestDocument ingestDocument) throws Exception {
                     .append('}')
                 .append('>');
         String dynamicIndexName  = builder.toString();
-        ingestDocument.setFieldValue(IngestDocument.MetaData.INDEX.getFieldName(), dynamicIndexName);
+        ingestDocument.setFieldValue(IngestDocument.MetaData.INDEX.getFieldName(), ValueSource.wrap(dynamicIndexName, scriptService));
```

Comment:
I think it is better that the `indexNamePrefix`, `dateRounding` and `indexNameFormat` become `TemplateScript.Factory` typed fields, like in other processor implementations. ValueSource is more meant to template generic values and in this case we know the templated values are always strings. Also then it is not needed to pass down `ScriptService` into the processor implementation and then the change in `IngestDocument` is not needed too.

Reply:
@martijnvg - thanks for the review !  the changes requested are at https://github.com/elastic/elasticsearch/pull/31841/commits/f9a6ec1e18c054b2184f03b6cfd429a940d75f1c.  This PR should be ready for re-review.